LFS-209: Linking between forms and subjects

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-209)

**To test**: Create a new subject from the user dashboard. It should allow you to select subjects. In addition, clicking the subject's name in a Form will allow you to change who that subject refers to. It should no longer be possible to create Forms with no subject.

OrdiNeu committed:
LFS-209: Linking between forms and subjects Allow the new form dialog to create new subjects

OrdiNeu committed:
LFS-209: Linking between forms and subjects Allow the new form dialog to create new subjects

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Move the subject movement to the form page

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Move the subject movement to the form page

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Allow subject filtering.
Improve responsiveness via UI improvements.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Allow subject filtering.
Improve responsiveness via UI improvements.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Code cleanup

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Code cleanup

sdumitriu commented:
Can you rename the commit messages please?

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Add button on the subject switching dialog

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Add button on the subject switching dialog

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Remove the ability for the user to create blank subjects

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Remove the ability for the user to create blank subjects

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Fix support for showing additional subjects.
Change the icon of subjects to AssignmentInd

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Fix support for showing additional subjects.
Change the icon of subjects to AssignmentInd

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Turn the dialog into a multi-stage product

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Turn the dialog into a multi-stage product

OrdiNeu committed:
[misc] Add a <!doctype> to the header of our web page.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Refactor everything to use material-tables instead.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Refactor everything to use material-tables instead.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Adjust style of the subject selector, and skip the questionnaire
selecting if presetPath is provided.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Adjust style of the subject selector, and skip the questionnaire
selecting if presetPath is provided.

sashaandjic commented:
Testing instance is running on http://172.20.4.76:8000/

It looks much better! Here are some problems that I observed:
- subject selection dialog always shows 1-1 of 1 
![image](https://user-images.githubusercontent.com/23243244/78329496-dbe05600-754f-11ea-9944-7cf0e7226d9f.png)
- when I create a new subject, the subject list does not get refreshed
- I can create multiple subject with the same name
- I can create multiple forms of the same type by using the same subject 
![image](https://user-images.githubusercontent.com/23243244/78329851-ad16af80-7550-11ea-8c6c-50e65340b44f.png)

IntegralProgrammer commented:
I have found a bug similar to what @sashaandjic described as "when I create a new subject, the subject list does not get refreshed". Here are the steps to reproduce it:

1. Checkout the `LFS-209` branch (`git checkout LFS-209`)
2. Build it (`mvn clean install`)
3. Start it (`java -jar distribution/target/lfs-0.1-SNAPSHOT.jar`)
4. Load `http://localhost:8080` and login as `admin`:`admin`
5. You should at this point be seeing an empty "User dashboard"
6. Create a new "Demographics" form
7. Click the `+` icon to create a new subject, name the subject `PatientX`, click checkmark
8. Observe that the subject list is empty ("No records to display")
9. Click the `+` icon to create a new subject, name the subject `PatientX`, click checkmark
10. This time, `PatientX` is now present in the list
11. Click `PatientX` in the list to select it.
12. Click the *Create* button
13. Fill out the demographics form and click the *Save* button
14. Now, click *Subjects* on the pane on the left side of the screen
15. `PatientX` is present twice - the older version linking to the newly completed demographics form, the newer version linking to nothing.

IntegralProgrammer commented:
I have found another bug, this time regarding the *Previous Page* and *Next Page* arrows in the *Select a subject* menu. Here are the steps to reproduce it:

1. Checkout the `LFS-209` branch (`git checkout LFS-209`)
2. Build it (`mvn clean install`)
3. Start it (`java -jar distribution/target/lfs-0.1-SNAPSHOT.jar`)
4. Load `http://localhost:8080` and login as `admin`:`admin`
5. You should at this point be seeing an empty "User dashboard"
6. Create a new "Demographics" form
7. Add 10 subjects (`patient0`, `patient1`, ... `patient9`)
8. The bottom of the *Select a subject* window shows `1-2 of 2` and clicking the *left* (*Previous Page*) or *right* (*Next Page*) button does not advance or retract the page. The only way that all 10 entries can be seen is by selecting *10 rows* or *20 rows* from the drop-down menu.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Remove debug code.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Remove debug code.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Remove error messages once the user selects a valid subject.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Remove error messages once the user selects a valid subject.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Fix material-ui handling of displaying new subjects, fixing a bug where
a new user  was not immediately selectable.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Fix material-ui handling of displaying new subjects, fixing a bug where
a new user  was not immediately selectable.

IntegralProgrammer commented:
Strictly speaking, this is not a bug, but I find the following to be different from what user expectations would most likely be:

1. Checkout the `LFS-209` branch (`git checkout LFS-209`)
2. Build it (`mvn clean install`)
3. Start it (`java -jar distribution/target/lfs-0.1-SNAPSHOT.jar`)
4. Load `http://localhost:8080` and login as `admin`:`admin`
5. You should at this point be seeing an empty "User dashboard"
6. Create a new "Demographics" form
7. Associate this new form with a new subject which we will call `PatientZ`
8. Fill out the form and *save* it.
9. Return to the *User dashboard* by clicking *User dashboard* on the top-left corner of the page.
10. Open the newly created form by clicking its link in the *Demographics Data* card
11. Click the large blue `PatientZ` title
12. Associate this form with a new subject, `PatientW`
13. Click *Subjects* in the pane on the left side of the screen
14. Click the link for *PatientW* and find that it is empty

In summary, what I suggest is that as soon as the *subject* associated with a *form* is changed, said *form* would be automatically saved, instead of relying on the user to manually click the *Save* button. Would you agree @marta- ?

OrdiNeu commented:
>     I have found another bug, this time regarding the Previous Page and Next Page arrows in the Select a subject menu.

>     * subject selection dialog always shows 1-1 of 1 
>     * when I create a new subject, the subject list does not get refreshed
>     * when deleting a subject is it possible to use smaller red font as one circled in blue? can we also include the subject name that the user is about to delete in the question?
>     * text circled in blue "Please select a subject" should clear up once newly created subject is either saved or canceled. 

These issues have now been fixed

OrdiNeu commented:
> I have found a bug similar to what @sashaandjic described as "when I create a new subject, the subject list does not get refreshed". Here are the steps to reproduce it:

This should now be fixed. Subjects are no longer located at their UUID, but instead at `/Subjects/<identifier>`

IntegralProgrammer commented:
@OrdiNeu 

- I have re-tested the issue with *Previous Page* and *Next Page* buttons not working properly in "Select a subject" menus and this has been fixed :heavy_check_mark: 

- Similarly, the issue with the "Select a subject" menu always displaying `1-1 of 1` has also been fixed :heavy_check_mark: 

- The issue with the subject list not being refreshed after a new subject has been created has been partially fixed. Sometimes the menu auto-reloads after creating a new subject while other times it does not. Possibly, this is the result of a data race?

- The issue of switching to a smaller red font has been fixed :heavy_check_mark: 

- The "Please select a subject" message no longer appears. It this okay?

sashaandjic commented:
With the current workflow this error message is no longer needed
![image](https://user-images.githubusercontent.com/23243244/79134622-ff8b7380-7d7b-11ea-8ab8-650aba702c87.png)

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Fix the data race condition where newly created subjects are not
immediately visible after creation.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Fix the data race condition where newly created subjects are not
immediately visible after creation.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Remove error messages when going backwards through the menu.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Remove error messages when going backwards through the menu.

OrdiNeu commented:
>     subject list not being refreshed after a new subject has been created has been partially fixed. Sometimes the menu auto-reloads after creating a new subject while other times it does not. Possibly, this is the result of a data race?

Fixed in `f3514bc` ✔ 

>     * The "Please select a subject" message no longer appears. It this okay?
>     With the current workflow this error message is no longer needed

Fixed in `8f7869b `. The message was only supposed to show up if the user was in error, so it not appearing until after they try to go back should be fine. ✔

veronikaslc commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -0,0 +1,298 @@
+            }
+          }
+        }}
+        onRowClick={(event, rowData) => {onSelect(rowData); console.log(rowData)}}
```
We don't need `console.log` after debugging

veronikaslc commented:
There is a problem: Say, a form had one subject, and then it was changed to the other subject. Clicking the subject's name in a Form allows you to delete current subject. Clicking on Confirm  button will throw error and won't close the dialog. Clicking away will close dialog. The form would be silently re-assigned to the subject is previously had.

veronikaslc commented:
On clicking create button for creating new subject autofocus should be set to the subject input.

veronikaslc commented:
Click on `Delete` trash can button in subject dialog should "unselect" any row previously selected. Also it should disable `Confirm` button.

veronikaslc commented:
Not a show-stopper, but abit counter-intuitive button tooltip for confirmation of a delete action.
![image](https://user-images.githubusercontent.com/4371097/79158835-05497f00-7da5-11ea-8807-7e9899e8b42f.png)

veronikaslc commented on modules/data-entry/src/main/frontend/src/dataHomepage/NewFormDialog.jsx:
```diff
@@ -20,31 +20,65 @@ import React, { useState } from "react";
+import { CircularProgress, Dialog, DialogContent, DialogTitle, Fab, Grid, List, DialogActions, Button} from "@material-ui/core";
+import { ListItemText, Tooltip, Typography, withStyles } from "@material-ui/core";
 import AddIcon from "@material-ui/icons/Add";
+import MaterialTable from "material-table";
```
Do you use it here?

veronikaslc commented on modules/data-entry/src/main/frontend/src/dataHomepage/NewFormDialog.jsx:
```diff
@@ -20,31 +20,65 @@ import React, { useState } from "react";
 
-import { Avatar, CircularProgress, Dialog, DialogContent, DialogTitle, Fab, List, ListItem, ListItemAvatar, ListItemText, Tooltip, Typography, withStyles } from "@material-ui/core";
-import AssignmentIcon from "@material-ui/icons/Assignment";
+import { CircularProgress, Dialog, DialogContent, DialogTitle, Fab, Grid, List, DialogActions, Button} from "@material-ui/core";
```
`Grid` is not used

IntegralProgrammer commented:
Re-tested and issue with *Subject list* not always being automatically refreshed after a new subject is added has been fixed :heavy_check_mark:

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Remove debugging code.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Remove debugging code.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Disallow the user from selecting a deleted user when creating a new
form.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Disallow the user from selecting a deleted user when creating a new
form.

IntegralProgrammer commented:
@OrdiNeu 

Commit `633abb1` is good as it denies a user from deleting a subject that is currently assigned to the given form, however, it would be nice if an error were to be displayed or the *trash can* icon could simply be omitted (or grayed out) for the subject identifier currently assigned to the given form.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Remove unused imports.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Remove unused imports.

sashaandjic commented:
retested. found the following:
- when creating new subject move focus to the subject line
- when creating a subject that already exist, it appears as it is allowed - however it does not get saved. user should be warned about the duplicate entry
- when attempting to delete a subject that is already in use, display a message teling user why they cannot delete it

sashaandjic commented:
Cannot search forms or subjects by subject

OrdiNeu commented:
> Not a show-stopper, but abit counter-intuitive button tooltip for confirmation of a delete action.

I can't seem to edit that one without also editing the tooltip when you save the edits to a row -- any ideas on a wording that works for both cases?

> Commit 633abb1 is good as it denies a user from deleting a subject that is currently assigned to the given form, however, it would be nice if an error were to be displayed or the trash can icon could simply be omitted (or grayed out) for the subject identifier currently assigned to the given form (or any other form that uses it).

I could maybe do this, if I disable the entire row (so disable editing the currently selected subject).

> Cannot search forms or subjects by subject

This is functionality added in #188

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Warn users about duplicate entries.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Warn users about duplicate entries.

OrdiNeu commented:
> when creating a subject that already exist, it appears as it is allowed - however it does not get saved. user should be warned about the duplicate entry 

Fixed in `e355f69` ✔

sashaandjic commented:
When I tried to save an empty subject I ended up with a blank screen. 
![image](https://user-images.githubusercontent.com/23243244/79272641-bcaac800-7e6f-11ea-9d15-8acafcb003e0.png)

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Crashfix when creating a blank subject.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Crashfix when creating a blank subject.

OrdiNeu commented:
> When I tried to save an empty subject I ended up with a blank screen.
Fixed in `f613048` ✔

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Move the creation of a new subject to a popover

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Move the creation of a new subject to a popover

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Make the ability to add/delete subjects an optional part of
SubjectSelector
Remove obsolete code.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Make the ability to add/delete subjects an optional part of
SubjectSelector
Remove obsolete code.

sashaandjic commented:
- problem with zoom above 80%
![image](https://user-images.githubusercontent.com/23243244/79945952-f9149000-843c-11ea-8505-f4a1126e9814.png)
- CREATE bttn should be blue; when creatinh a new form from Forms CANCEL button has BACK as a label; 
![image](https://user-images.githubusercontent.com/23243244/79947391-e8194e00-843f-11ea-95cb-2baae3e0077b.png)
![image](https://user-images.githubusercontent.com/23243244/79946285-a7203a00-843d-11ea-91ac-274dc788efa7.png)
![image](https://user-images.githubusercontent.com/23243244/79947610-3b8b9c00-8440-11ea-8bd4-679960cd044c.png)

- I would prefer having **+** to CREATE NEW SUBJECT button 
- creating duplicate subject isn't allowed, however there should be an error msg informing user that the subject already exist
- when changing subject on the form user can select from the list; why wouldn't we allow creation of the new subject
![image](https://user-images.githubusercontent.com/23243244/79946922-e69b5600-843e-11ea-93dd-96eb3a050a88.png)

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Rearrange UI, add new subject functionality to Form, and remove dead
code.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Rearrange UI, add new subject functionality to Form, and remove dead
code.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Fix error reporting

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Fix error reporting

OrdiNeu commented:
>     * I would prefer having "**+**" bttn at the top to CREATE NEW SUBJECT button at the bottom
> 
>     * focus should always be on the CREATE button (this is probably out of scope of this task) 

Everything except these two have been addressed in the latest commits.

sashaandjic commented:
The form looks OK now. The BACK button instead of CANCEL is still appearing:
![image](https://user-images.githubusercontent.com/23243244/80061783-2aa25f80-8500-11ea-891f-4b4c16852117.png)
I would use this form instead of one below (the only diff between them is missling CANCEL bttn which is a problem)
![image](https://user-images.githubusercontent.com/23243244/80061840-47d72e00-8500-11ea-951c-589bed784f48.png)

There is a problem with special characters when creating a new subject:

veronikaslc commented on modules/data-entry/src/main/frontend/src/dataHomepage/NewFormDialog.jsx:
```diff
@@ -55,62 +89,203 @@ function NewFormDialog(props) {
+        <DialogContent dividers className={classes.NewFormDialog}>
+          {error && <Typography color='error'>{error}</Typography>}
+          <Input
+            value={newSubjectName}
```
It would be nice to have `autoFocus="true"` here so the input field is focused during the first mount.

veronikaslc commented:
There is no `Cancel` button. I suppose this functionality should b presented.
![image](https://user-images.githubusercontent.com/4371097/80141832-2c0f6e80-8578-11ea-82b2-547539a9c094.png)

veronikaslc commented:
Its impossible to change subject to another existing one
![image](https://user-images.githubusercontent.com/4371097/80145494-3fbdd380-857e-11ea-9754-2c91060828a4.png)

veronikaslc commented:
Creating a duplicate subject disables all buttons in the dialog
![image](https://user-images.githubusercontent.com/4371097/80155958-c9769c80-8590-11ea-9432-0ebcbfc0eba1.png)

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Refactor subjects back to UUIDs, and fix the UI for the switch subjects
page.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Refactor subjects back to UUIDs, and fix the UI for the switch subjects
page.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Various bugfixes with the submission of new subjects on the forms page.

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Various bugfixes with the submission of new subjects on the forms page.

sashaandjic reviewed (approved):
**TESTED! GOOD!**
Found some very minor things that will be logged separately as they should not postpone this PR.

marta- commented:
Looks good enough to be merged. Some very quick ui improvements :
* "Create subject" should be labeled "New subject" and should have `color="secondary"`
* "Cancel" and "Back" buttons should have `color="default"` in all dialogs

sashaandjic commented:
> * "Create subject" should be labeled "New subject" and should have `color="secondary"`

Should all CREATE bttn labels be replaced with NEW?
@marta-

marta- commented:
> Should all CREATE bttn labels be replaced with NEW?

I'd say no. While the mismatch may raise consistency concerns, the buttons actually do different things: one pops yet another dialog for entering the new subject, while the other one finishes the process and takes the user to the form.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/NewFormDialog.jsx:
```diff
@@ -55,62 +89,203 @@ function NewFormDialog(props) {
+              onClick={() => { setNewSubjectPopperOpen(true); setError(); }}
+              className={classes.createNewSubjectButton}
+              >
+              Create Subject
```
```suggestion
              color="secondary"
              onClick={() => { setNewSubjectPopperOpen(true); setError(); }}
              className={classes.createNewSubjectButton}
              >
              New subject
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/NewFormDialog.jsx:
```diff
@@ -55,62 +89,203 @@ function NewFormDialog(props) {
+          }
+          <Button
+            variant="contained"
+            color="secondary"
```
```suggestion
            color="default"
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/NewFormDialog.jsx:
```diff
@@ -55,62 +89,203 @@ function NewFormDialog(props) {
+          <Button
+            onClick={() => {setNewSubjectPopperOpen(false); setError();}}
+            variant="contained"
+            color="secondary"
```
```suggestion
            color="default"
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -0,0 +1,388 @@
+        <Button
+          onClick={closeNewSubjectPopper}
+          variant="contained"
+          color="secondary"
```
```suggestion
          color="default"
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -0,0 +1,388 @@
+      <DialogActions>
+        <Button
+          variant="contained"
+          color="primary"
+          onClick={() => { setNewSubjectPopperOpen(true); }}
+          className={classes.createNewSubjectButton}
+          >
+          Create Subject
```
```suggestion
          color="secondary"
          onClick={() => { setNewSubjectPopperOpen(true); }}
          className={classes.createNewSubjectButton}
          >
          New subject
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -0,0 +1,388 @@
+        <Button
+          onClick={onClose}
+          variant="contained"
+          color="secondary"
```
```suggestion
          color="default"
```

OrdiNeu committed:
LFS-209: Linking between forms and subjects

Apply suggestions from code review by @marta-

Co-Authored-By: Marta Girdea <marta.girdea@gmail.com>

OrdiNeu committed:
LFS-209: Linking between forms and subjects

Apply suggestions from code review by @marta-

Co-Authored-By: Marta Girdea <marta.girdea@gmail.com>

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/NewFormDialog.jsx:
```diff
@@ -55,62 +89,203 @@ function NewFormDialog(props) {
-    // Determine what questionnaires are available
+    // Determine what questionnaires and subjects are available
+    if (!initialized) {
+      fetchAndPopulate("select * from [lfs:Questionnaire]", setQuestionnaires);
```
This only gets the first 10 questionnaires. Add a bigger `limit`.

OrdiNeu replied:
What limit is reasonable here?

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -0,0 +1,388 @@
+            addTooltip: "Add a new subject",
+            editRow: {
+              /* NB: We can't escape the h6 placed around the delete text, so we instead override the style */
+              deleteText: <span className={classes.deleteText}>Are you sure you want to delete this row?</span>
```
Is this still needed?

OrdiNeu replied:
If `onAllowDelete` is enabled, this will be displayed. It currently isn't used, but I figured it might be useful if we want to reuse this table in a context where these options are enabled.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -0,0 +1,388 @@
+ * Component that displays the list of subjects.
+ *
+ * @example
+ * <Form id="9399ca39-ab9a-4db4-bf95-7760045945fe"/>
```
Wrong example.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -0,0 +1,388 @@
+ * <Form id="9399ca39-ab9a-4db4-bf95-7760045945fe"/>
+ *
+ * @param {disabled} bool whether selections should be disabled on this element
+ * @param {onDelete} func Callback for the deletion of a subject. The only parameter is the subject deleted.
```
Leftover?

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -0,0 +1,388 @@
+          >
+          Cancel
+        </Button>
+        <Button
```
This should respond to pressing Enter.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -0,0 +1,388 @@
+ * @param {onSelect} func Callback for when the user selects a subject.
+ * @param {selectedSubject} object The currently selected subject
+ * @param {setSubjects} func A callback for setting the currently available subjects. The parameter is a list of subjects
+ * @param {subjects} array A list of (potentially filtered) subjects that the user has available
```
The documentation needs to be updated to reflect the actual props.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -0,0 +1,388 @@
+              )
+          }
+        }
+        editable={{
```
Is this still needed?

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -0,0 +1,388 @@
+                )
+              );
+          } : undefined),
+          onRowDelete: (allowDeleteSubjects ? oldData => {
```
Is this still needed?

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Remove unused code, cleanup documentation

OrdiNeu committed:
LFS-209: Linking between forms and subjects
Add autofocus to the NewFormDialog
Submit the new subject when the user presses enter.

sashaandjic reviewed (approved):
re-tested
